openshift/dynamic/apply.py

```diff
@@ -70,6 +114,42 @@ def dict_merge(a, b):
     return result
 
 
+def list_to_dict(lst, key, position):
+    result = OrderedDict()
+    for item in lst:
+        try:
+            result[item[key]] = item
+        except KeyError:
+            raise ApplyException("Expected key '%s' not found in position %s" % (key, position))
+    return result
+
+
+# list_merge applies a strategic merge to a set of lists if the patchMergeKey is known
+# each item in the list is compared based on the patchMergeKey - if two values with the
+# same patchMergeKey differ, we take the keys that are in last applied, compare the
+# actual and desired for those keys, and update if any differ
+def list_merge(last_applied, actual, desired, position):
+    result = list()
+    if position in STRATEGIC_MERGE_PATCH_KEYS and last_applied:
+        patch_merge_key = STRATEGIC_MERGE_PATCH_KEYS[position]
+        last_applied_dict = list_to_dict(last_applied, patch_merge_key, position)
+        actual_dict = list_to_dict(actual, patch_merge_key, position)
+        desired_dict = list_to_dict(desired, patch_merge_key, position)
+        for key in desired_dict:
+            if key not in actual_dict or key not in last_applied_dict:
+                result.append(desired_dict[key])
+            else:
+                # This won't delete any keys that are in last_applied but not in desired
+                if any([desired_dict[key].get(entry) != actual_dict[key].get(entry)
```

Comment:
Don't we want to delete keys that are in last applied but not desired? I thought that's how we compute deletions

Reply:
You're right, I was thinking about this a little narrowly (e.g. envvars and ports which are unlikely to lose keys - whereas containers might lose a key such as command or args if you decide to override and then go back to defaults)